src/util/mod.rs

```diff
@@ -53,6 +55,10 @@ mod thread_metrics;
 
 pub const NO_LIMIT: u64 = u64::MAX;
 
+pub trait AssertSend: Send {}
```

Comment:
Can you explain why do we need to use an Assert trait here?

Reply:
Trait `AssertSend/Sync` provides a utility (for this reason it is stay in `util`) to assert that a specific structure is `Send/Sync`, thanks to the trait bound functionality provided by Rust compiler. For example, when we have a struct `Foo` and want to make sure that it is `Send`, we can write `impl AssertSend for Foo {}`. Then at the compile time, the compiler will help us check that `Foo` is `Send` and reject with errors if not.

The benifit of `AssertSend/Sync` is, when someone modifies `Foo` later and breaks the `Send` or `Sync` by mistake, the compiler will catch the mistake **in the scope of the `Foo` itself**, instead of bubbling it up to some structure else that used the `Foo` and may be reporting another error.

Reply:
em, why not using Send directly?

Reply:
Because it just can't achieve "assert“. For example, `impl Send for Foo` means mark `Foo` as `Send`, not check whether it is `Send`.

Reply:
This "trick" in fact is borrowed from futures-cpupool: https://github.com/alexcrichton/futures-rs/blob/master/futures-cpupool/src/lib.rs#L95